src/h5web/vis-packs/core/complex/hooks.ts

```diff
@@ -0,0 +1,74 @@
+import ndarray from 'ndarray';
+import { useMemo } from 'react';
+import type { DimensionMapping } from '../../../dimension-mapper/models';
+import type { H5WebComplex } from '../../../providers/models';
+import { useMappedArray } from '../hooks';
+import { Bounds, Domain, ScaleType } from '../models';
+import { getDomainValidForScale, getNewBounds } from '../utils';
+
+export function usePhaseAmplitude(
+  value: H5WebComplex[],
+  dims: number[],
+  dimMapping: DimensionMapping,
+  scaleType: ScaleType = ScaleType.Linear
+): {
+  phaseArray: ndarray;
+  phaseDomain: Domain | undefined;
+  amplitudeArray: ndarray;
+  amplitudeDomain: Domain | undefined;
+} {
+  const [mappedArray] = useMappedArray(value, dims, dimMapping);
+
+  const [
+    phaseArray,
+    phaseBounds,
+    amplitudeArray,
+    amplitudeBounds,
+  ] = useMemo(() => {
+    const phaseValues: number[] = [];
+    const amplitudeValues: number[] = [];
+
+    let phaseBounds: Bounds = {
+      min: Infinity,
+      max: -Infinity,
+      positiveMin: Infinity,
+    };
+    let amplitudeBounds: Bounds = {
+      min: Infinity,
+      max: -Infinity,
+      positiveMin: Infinity,
+    };
+
+    (mappedArray.data as H5WebComplex[]).forEach(([real, imag]) => {
+      const phase = Math.atan2(imag, real);
+      phaseBounds = getNewBounds(phaseBounds, phase);
+      phaseValues.push(phase);
+
+      const amplitude = Math.sqrt(real ** 2 + imag ** 2);
+      amplitudeBounds = getNewBounds(amplitudeBounds, amplitude);
+      amplitudeValues.push(amplitude);
+    });
+
+    return [
+      ndarray(phaseValues, mappedArray.shape),
+      phaseBounds,
+      ndarray(amplitudeValues, mappedArray.shape),
+      amplitudeBounds,
+    ];
+  }, [mappedArray]);
+
+  const [phaseDomain, amplitudeDomain] = useMemo(
+    () => [
+      getDomainValidForScale(phaseBounds, scaleType),
+      getDomainValidForScale(amplitudeBounds, scaleType),
+    ],
+    [amplitudeBounds, phaseBounds, scaleType]
+  );
```

Comment:
The fact that you've separated these two `useMemo` makes me wonder whether there's an improvement to be made to `useDomain` as well. It seems we must be recomputing the domain by looping over the entire data array every time the scale changes, when this could be avoided.

Reply:
Yep, testing with `entry_0000/0_measurement/images` confirms that switching the scale blocks the UI, indicating that the domain is being recomputed from scratch.